Fix some of the issues with the migration service and the migration progress dialog

oakkitten opened the pull request:
**Edit**: comment updated to reflect changes brought in in c7330ea.  
**Edit**: comment updated to reflect changes pushed on April 15.


This is built on top of #13219. Closes #13219, closes #13299. ~~Draft PR because one TODO in `MigrationService` must be addressed.~~ Ready to be reviewed.

## Purpose / Description

See discussion in #13219.

## Approach

The service was modified to emit a `StateFlow` of `Progress`, and it is collected in both the service (for the notification), the deck picker (to set the ongoing-migration toolbar button) and in the progress dialog it opens. The dialog no longer depends on activity, or a view model; it reads the flow directly from the service. In my testing, binding to the service happens *very* fast, and no optimizations that I mentioned earlier are necessary.

~~Warning: binding to the service will create the service. For simplicity, UI elements do not check whether the service has been started or not, but bind to it unconditionally. This allows UI elements to fetch status information directly from the service, even after the service has ran to completion. There's one TODO on `onCreate` that ***MUST*** be addressed in order for this to work. An alternative approach would be to use ViewModels to keep track of the last migration status. Yet another alternative would be making a singleton flow, but I didn't want to change the logic too drastically there as it may disrupt testing.~~

Note: the service will remain active until the last client unbinds from it. ~~While the service is not stopped, its notification also cannot be removed.~~ Edit: see next comment.

~~Note: the indeterminate `ProgressBar` in the migration progress dialog has inadequate appearance.~~ Using `CircularProgressIndicator` for both the toolbar icon and the migration progress dialog now.

Note: `onStartCommand` is called twice, there's a safeguard but why is it called twice?

~~Note: there's a merge conflict that I am not sure about. Expediting moving of certain files sounds too complicated, can we instead simply try old location first, and then fall back to the new location?~~ I rebased onto current main while keeping most of the logic. I tested this on my device and it works, but this **really** needs to be looked at by someone else.

## How Has This Been Tested?

Tested manually on my device.

## Checklist
_Please, go through these checks before submitting the PR._

- [ ] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

oakkitten commented:
Updated to call `stopForeground`. Now on ≥ N on the completion of the migration the notification stays, *and* can be dismissed with swiping. More testing might be required to ascertain the behavior prior to N

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/services/MigrationService.kt:
```diff
@@ -16,262 +16,242 @@
-            val listener = MigrateUserDataProgressListener(this)
-            listener.initNotification(totalToTransfer)
+        lifecycleScope.launch(Dispatchers.IO) {
+            flowOfProgress.emit(Progress.CalculatingTransferSize)
```
`CalculatingTransferSize` is the default value. Unless you expect this `onStartCommand` to be executed multiple time on the same object, this line can probably be removed.

oakkitten replied:
I'm a bit unsure here. I mean yes, it _can_ be removed, but it's just so nice to have the first value emitted at the start of the block. I don't have a good argument for it—apart from it somewhat safeguarding from code changes, like if later `Progress.Unknown` is added as the default value or something—but it kind of feels right...

Arthur-Milchior replied:
Not really YAGNI, to safeguard from code change.
Anyway, it does not hurt, so I won't block on it

oakkitten replied:
Actually, I suppose the reason why it feels right for me is because it makes it easier to reason about the code *now*. You don't have to ask, what's the default value? And you don't have to ask, does this block only ever run once? Following the pattern of methods such as

https://github.com/ankidroid/Anki-Android/blob/a06e727f6094541b763e2dc18a300edb3e17fd26/AnkiDroid/src/main/java/com/ichi2/anki/ui/windows/managespace/ManageSpaceFragment.kt#L150-L157

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/services/MigrationService.kt:
```diff
@@ -16,262 +16,242 @@
+                flowOfProgress.emit(Progress.Success)
+
+                // TODO This is probably not supposed to be here
+                AnkiDroidApp.getSharedPrefs(this@MigrationService).edit {
```
I fell like it should be before a potentially blocking operation. If for some reason the flowOfProgress emit blocks, at least, preferences are correct.

oakkitten replied:
This is a bit of technicality, but I don't think `StateFlow.emit` can block, judging from the fact that `tryEmit` and `emit` have the same code. But anyway, it should not matter. There's a tiny chance that the app is killed at any time before writing the success state to preferences, a tiny chance but a chance nonetheless. Something we can't avoid. We should be able to gracefully recover from that, and if we do, it doesn't matter whether the `emit` call can take time or not.

oakkitten replied:
Actually I changed the order in my branch anyway because it's a good idea to finish up everything before calling any listeners, anyway. 👍

Oh and stopping the service will not actually stop it until all the clients unbind, anyway, IIRC. Also the service doesn't really have a lifecycle since it has no `onStart`/`onStop` so I kind of made a fake one here for easier reading.

BrayanDSO replied:
idk if it helps at all , but fwiw there is a LifecycleService class that can be used (reinforcing that I barely read this and I'm not sure at all if it will help, but I wanted to share the info)

oakkitten replied:
I tried putting in `LifecycleService` and it's, well. It does give you the lifecycle scope, but you have to call super in various methods, which is slightly annoying. Also it gives you a lifecycle, which I am not sure how much sense it makes as there's no `onStart` (in the same sense) and `onStop` in a service... And it requires another dependency as well. Not sure I likes

Anyway, I'll mark this as resolved as the initial issue has been resolved

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/services/MigrationService.kt:
```diff
@@ -16,262 +16,242 @@
+    override fun onBind(intent: Intent): IBinder = LocalBinder(this)
+}
+
+private fun Context.makeMigrationProgressNotification(progress: MigrationService.Progress): Notification {
```
I think we have generally avoided adding method to Context, unless they are very general and actually makes context more useful.
I'd consider moving context as an argument

oakkitten replied:
Why avoid? `Context` is such an ubiquitous thing in Android, it seems to me it should be a perfect candidate for extension functions. While this function is not general, it is `private` and does not pollute the namespace.

Arthur-Milchior replied:
I don't have a strong opinion myself. I recall that @david-allison objected when I was extending Context myself months ago. Maybe he will be able to answer once back.

Anyway, given it's private, I guess you're perfectly right it even matter less than usual, so I'll be fine with that too

Arthur-Milchior reviewed (commented):
Thanks for your work. I clearly admit the flow is easy to follow. 

You mention in the PR comment you tested manually. How? By forcing a migration on your phone?

Could you provide screenshot, it would help. Especially when you ention that indeterminate ProgressBar has inadequate appearance.

`onStartCommand` is called twice. I assume you mean this is something you remarked while manually testing the code. Could I suggest adding a break, so that maybe the debugger can show you two different paths, helping explaining the difference.  I assume it would be also nice to put the break on `startMigrateUserDataService` since it probably this method that is somehow called twice

Finally, I don't understand what conflict you are referring to in 
> Note: there's a conflict that I am not sure about. Expediting moving of certain files sounds too complicated, can we instead simply try old location first, and then fall back to the new location?

Are you referring to the fact that files can be preempted if they should be displayed? If so, I don't understand how this relate to the current PR

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt:
```diff
@@ -2894,29 +2876,15 @@ open class DeckPicker :
-                openUrl(R.string.link_scoped_storage_faq)
-            }
-        }
+        MigrationProgressDialogFragment().show(supportFragmentManager, "MigrationProgressDialogFragment")
```
I would avoid creating a constant here. A literal can even serve as an indicator that the tag is not used elsewhere! If it is ever used elsewhere, extracting it into a constant will be a very simple change. (I'm a fan of the YAGNI principle)

Arthur-Milchior replied:
I didn't know this principle or acronym. Thanks for teaching.
Agreed, makes sense now that you explain it.

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/services/MigrationService.kt:
```diff
@@ -16,262 +16,242 @@
-            /** @see notificationUpdater for where this is used */
-            currentProgress += value ?: 0
-        }
+    interface Progress {
```
I'm not a fan of comments that don't add anything to the identifier, `MigrationService.Progress`. Perhaps rename it to `MigrationService.ProgressState`? I'm not sure if this is more understandable...

Arthur-Milchior replied:
I know you don't like, so I really tried to only ask you to do it when I had trouble guessing the meaning of the type.

ProgressState is better as far as I'm concerned, because at least it's clear that you can't use it to do any progress, that you only use it to represent a state. Which, I believe, was my main concern

oakkitten replied:
I thought about this a little more and now I think `ProgressState` doesn't help much. You can ask both “what's the progress of the migration?” and “what's the state of the migration” and you'd be meaning the same thing, so “a state of progress“ is a bit of a tautology. I suppose it only feels right because it's common to suffix stuff with “state”...

oakkitten commented:
Updated to include fixes for review comments marked with 🚀.

> You mention in the PR comment you tested manually. How? By forcing a migration on your phone?

Yes. I hope we make a way to simplify this process.

> Could you provide screenshot, it would help. Especially when you ention that indeterminate ProgressBar has inadequate appearance.

https://user-images.githubusercontent.com/1710718/225995637-1d712993-bb1d-4b40-b3b4-bb4ae27a8027.mp4

  
> `onStartCommand` is called twice. I assume you mean this is something you remarked while manually testing the code. Could I suggest adding a break, so that maybe the debugger can show you two different paths, helping explaining the difference. I assume it would be also nice to put the break on `startMigrateUserDataService` since it probably this method that is somehow called twice

Yes, while manually testing. This shouldn't be a problem, as we have a safeguard, and yes this can be easily investigated, I was just lazy since my changes to the code didn't affect that part. Anyone can investigate!

> Finally, I don't understand what conflict you are referring to in

Sorry, meant merge conflict, edited.

oakkitten commented:
I'm currently working on rebasing this on `main` and making it properly work with #13335 and the two issues mentioned above.

oakkitten commented:
Rebased onto current main and updated to work with #13335. Using `CircularProgressIndicator` for both the ongoing-migration button and the dialog.

https://user-images.githubusercontent.com/1710718/227634058-732061a7-37b7-4912-9307-e699f3376a6d.mp4

There are still some TODOs in `MigrationService` that ought to be addressed before this is merged. These are related to the inner workings of `MigrateUserData` which I dare not touch.

I usually re-re-review my PRs over the course of several days, or even weeks, however, looking forward to these fixes, which may or may not require considerable changes, and also because we are pressed for time a bit, I decided to publish this now. So when reviewing, please be thorough. Also, for some reason the linter is broken on my computer so I had to disable git hooks, so expect that this includes some lint errors that should get fixed.

oakkitten commented:
Updated to mark the TODOs as to be looked at BEFORE-MERGE or BEFORE-RELEASE. See the commit message for more info.

Also I want to note that this PR removes any rate limiting on notification updates. I simply don't think this is necessary in practice with the way `MigrateUserData` reports progress—it is not fast enough to cause any issues. Besides, the system applies its own variable rate limiting to notification updates, and specifically makes an exception for updates with a 100% completeness.

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/services/MigrationService.kt:
```diff
@@ -16,260 +16,169 @@
-            runOnUiThread {
-                UIUtils.showThemedToast(context, message, true)
-            }
+        // TODO BEFORE-MERGE `MigrateUserData.createInstance` is doing I/O and can throw I/O errors.
+        //   Figure out what's going on and move all I/O into a thread.
```
Can this be peeled off into a TODO so this code can get out, or does it hard block the merge?

Verifies that the source/destination of the migration are valid to continue

* Invalid Preferences: probably impossible - destination/source directory are badly defined
* Source/destination do not exist (or are non-directory objects)

oakkitten replied:
Doesn't block the merge. Not really an issue if the code is correct—will block the main thread for a bit but probably not for long, but it would be easier to understand the behavior and to make sure that the code is correct if `createInstance()` didn't do I/O. Up to you

david-allison replied:
Don't block the merge over it

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/services/MigrationService.kt:
```diff
@@ -16,260 +16,169 @@
-            }
+        // TODO BEFORE-MERGE `MigrateUserData.createInstance` is doing I/O and can throw I/O errors.
+        //   Figure out what's going on and move all I/O into a thread.
+        // TODO BEFORE-MERGE Previous code re-threw MigrateUserData.MissingDirectoryException.
+        //   Figure out why.
```
Specific case where the user would have a collection on removable storage, and the removable storage is unmounted/removed.

This is a case where we would want to notify the user that the service failed to launch.

oakkitten replied:
Current code will emit `Progress.Failure` on all exceptions so I suppose the UI is able to deal with the problem. I suppose we do need some UI to show the errors though. These are shown in a very primitive way in the exception, but as I think I've noted in #13131, we seem to be not dealing with errors there.

If `MigrateUserData.MissingDirectoryException` needs special treatment, suggest creating another issue

david-allison replied:
Sounds good. Will do

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/services/MigrationService.kt:
```diff
@@ -16,260 +16,169 @@
-            currentProgress += value ?: 0
-        }
+    // TODO BEFORE-MERGE! Emit the according `flowOfProgress` here.
+    //   I think it would simplify things if we could always bind to the service,
```
Sounds good

oakkitten replied:
I was hoping you'd do this.

(In real world, in current implementation this probably only matters when the deck picker is recreated while showing the progress dialog. The dialog after recreation will try to connect to the service, and if at some time before the migration finishes, and the service is destroyed, the dialog might show the wrong state.

What I would actually want to see here would be moving `Progress` and the related to `MigrateUserData`, since it doesn't really belong here. But probably something like this will be enough:

```kt
if (migration has completed) tryEmit(Done)
```

Alternatively, `flowOfProgress` could be a global (companion object?) variable.)

david-allison replied:
You're in the best position to do this.

mikehardy replied:
Ok - so if this is todo-before-merge, do you have any plans for this @oakkitten ? Ideally whatever is **the minimal viable solution** is the thing that happens. Just not crashing, and not actually blocking the migration, would be acceptable as a first cut to get this merged

oakkitten replied:
I can attempt to fix this and deal with anything else related to `MigrateUserData`. But please note that due to my poor understanding of the inner workings of it, I might not be able to guarantee code correctness to the best of my ability. As this PR seems to be important, I'll try to set everything else aside for it, but I won't have anything until this weekend.

FWIW, the current code does not crash, the issue is that it currently introduces slightly wrong behavior. It's technically mergeable. But also it will probably fail some lint checks as my linter is (was?) broken, and also there are some failing tests for the reasons mentioned above.

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/services/MigrationService.kt:
```diff
@@ -16,260 +16,169 @@
-    inner class LocalBinder : Binder(), SimpleBinder<MigrationService> {
-        @Suppress("unused")
-        override fun getService(): MigrationService = this@MigrationService
+    // TODO BEFORE-RELEASE! This is inadequate, instead of calculating the remaining transfer size
+    //   every time migration is started,
+    //   we should be calculating this size only when starting migration for the first time,
+    //   and keeping track of the transferred and the remaining files.
```
https://github.com/orgs/ankidroid/projects/1/views/1?pane=issue&itemId=21353064

mikehardy replied:
okay, tracked externally

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/scopedstorage/migrateuserdata/MigrateUserData.kt:
```diff
@@ -529,14 +529,11 @@ open class MigrateUserData protected constructor(val source: Directory, val dest
             context.terminatedWith?.let { throw it }
-            val migrationFailedMessage = AnkiDroidApp.instance.getString(R.string.migration_failed_message)
-            throw AggregateException.raise(migrationFailedMessage, context.loggedExceptions)
+            throw AggregateException.raise("Migration failed", context.loggedExceptions)
```
This exception is raised with what seems to be a localized message. Must be a mistake?

david-allison replied:
You're correct

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/services/MigrationService.kt:
```diff
@@ -16,260 +16,169 @@
-            Timber.w(e, "Failed to get directory size")
-            null
-        }
+        // TODO BEFORE-MERGE Why the below line was necessary?
+        //   if (::migrateUserDataTask.isInitialized) { migrateUserDataTask.executor.terminate() }
```
Sorry, killing which thread? Who kills it? I'm not sure I understand

david-allison replied:
Previously we were running inside a thread inside `MigrationService`. Instead of killing the thread in `onDestroy()`, we called .terminate() on the executor to allow for it to exit gracefully.

oakkitten replied:
Ah I see. My question was more like, do we expect the service to ever be destroyed while the migration is running? As I understand, foreground services will never be stopped unless `stopSelf` is called, which we only do on migration completion.

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/services/MigrationService.kt:
```diff
@@ -16,260 +16,169 @@
-        }
-        isStarted = true
-        Timber.d("onStartCommand")
+                // TODO BEFORE-RELEASE Replace the toast with a dialog that is shown either
+                //   * now, if deck picker is shown,
+                //   * whenever the deck picker is shown, if not, perhaps even after app death.
+                if (progress is Progress.Success) {
+                    UIUtils.showThemedToast(this@MigrationService, R.string.migration__successful, true)
+                }
```
Yeah I was hoping #13131 will deal with this after this PR is merged. I collaborate on that one if need be

criticalAY replied:
We can show the toast as well as the dialog here? I knowingly didn't remove this toast in the STOR-05 PR

oakkitten replied:
I would personally remove the toast. I generally dislike seeing toasts from background applications—it's so often not clear where it comes from. The user is going to see the status in the notification, and also they will see a dialog when returning to the app.

criticalAY replied:
Ok so if we remove the toast then we would be left with 
- Notification
- Snackbar in case we have the application running
- A dialog which would be shown as the migration is completed or when the user resumes the application

oakkitten replied:
If the application is running, I'd show a dialog as well, not a snackbar. I just feel like this final message is kind of important

criticalAY replied:
Yes!! it is important agreed so dialog + notification then cool

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/services/ServiceUtils.kt:
```diff
@@ -0,0 +1,104 @@
+ *
+ *     val migrationService: MigrationService? by serviceWhileStartedOrNull()
+ *
+ * Note: binding to the service happens fast, but not immediately,
+ * so expect this property to be `null` when reading right after `onStart()`.
```
As I understand, binding to the service happens entirely on main thread. I'd assume web view also does a few main thread dispatches before it gets to loading files. I bet the service wins (and since we are talking one thread only, it should do that every time)

Either way, `ServiceConnection` had the same issue

oakkitten commented:
Rebased on main and updated to:
* resolve conflicts;
* fix tests;
* fix a small issue is `AbstractFlashcardViewer`; the previous code was trying to bind to the service unconditionally, and the new code should behave exactly as the one before this PR, only binding if the migration is in progress when the activity is started;
* remove/resolve/edit/add TODOs, mainly so that none block this PR—see individual commits for detail.

I usually spend quite a bit of time re-reviewing my own code. This time we are in a hurry so I am publishing this without paying attention to the detail as much as I'd like. Please review thoroughly. I sincerely hope that the reviewers test the code in a real-world setting.

david-allison commented:
@criticalAY tagging as you were requested as a reviewer in Discord

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/services/MigrationService.kt:
```diff
@@ -16,260 +16,157 @@
-            stopSelf()
-            return getRestartBehavior()
-        }
+                // TODO BEFORE-RELEASE Replace the toast with a dialog that is shown either
+                //   * now, if deck picker is shown,
+                //   * whenever the deck picker is shown, if not, perhaps even after app death.
```
https://github.com/orgs/ankidroid/projects/1/views/1?pane=issue&itemId=25747138

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/services/MigrationService.kt:
```diff
@@ -288,6 +185,85 @@ class MigrationService : Service() {
+            builder.setContentText(getString(R.string.migration__successful))
+        }
+
+        // TODO BEFORE-RELEASE Add a “Get help” button
+        is MigrationService.Progress.Failure -> {
```
https://github.com/orgs/ankidroid/projects/1/views/1?pane=issue&itemId=25747166

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/services/MigrationService.kt:
```diff
@@ -288,6 +185,85 @@ class MigrationService : Service() {
+        }
+
+        is MigrationService.Progress.Transferring -> {
+            val progressRatio = progress.transferredBytes.toFloat() / progress.totalBytes
```
`progress.totalBytes (getRemainingTransferSize)` can return 0

criticalAY replied:
`totalBytes` might be zero are we handling that somewhere?

david-allison commented on AnkiDroid/src/main/res/values/01-core.xml:
```diff
@@ -238,9 +238,12 @@
+    <string name="migration__transferred_x_of_y"
+        comment="The placeholders will be replaced with short localized byte amounts,
+        e.g. Transferred 10 kB of 5.2 MB">Transferred %1$s of %2$s</string>
+    <string name="migration__successful">Migration successful</string>
```
This still shouldn't need a key change

david-allison reviewed (changes requested):
Thanks!!!

One divide by zero issue I want to be paranoid with, the rest LGTM

criticalAY commented on AnkiDroid/src/main/res/layout/dialog_with_progress_indicator.xml:
Just a nit (upto you if you wanna implement it or not ) but we can remove the  `android:layout_gravity="center"` from the elements in the layout and use ` android:gravity="center"` in the layout itself as this layout is specific to the progress layout only

criticalAY commented on AnkiDroid/src/main/java/com/ichi2/anki/services/MigrationService.kt:
```diff
@@ -16,260 +16,157 @@
-        }
+                if (progress is Progress.Done) {
+                    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.N) {
+                        @Suppress("DEPRECATION")
```
Not so important right now but would like this deprecation to be handled in future this is not a blocker for this PR so good to go.

oakkitten replied:
This method is only used for API before N, when it's not deprecated. This `if` block is similar to what `Compat` classes do, I just think that using an `if` block instead of a bunch of classes and an interface simplifies things, as you are seeing everything in one place.

criticalAY commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/MigrationProgressDialogFragment.kt:
```diff
@@ -0,0 +1,97 @@
+                }
+
+                is MigrationService.Progress.Transferring -> {
+                    val progressRatio = progress.transferredBytes.toFloat() / progress.totalBytes
```
Same divide by zero. I mentioned that earlier that it is a rare case that the user is going to be able to reach this action to display the dialog so fast in case its zero but just to be sure we can consider placing a check here

criticalAY commented on AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt:
```diff
@@ -596,28 +590,86 @@ open class DeckPicker :
+                // TODO BEFORE-RELEASE Perhaps handle the cases of success & failure differently?
+                is MigrationService.Progress.Done -> {
+                    updateMenuState()
+                    updateMenuFromState(menu)
```
Can we invalidate the menu on the migration completion I am not sure I am facing this issue on the emulator right now but as soon as the migration was completed the menu state was not refreshed and required a change of activity i.e. going back to another screen and coming back to the deck picker

oakkitten replied:
Hm, this code should be updating the menu though? This is what I see:

https://user-images.githubusercontent.com/1710718/232296645-16c94791-6a53-4798-a1b5-d1d702e2c507.mp4

criticalAY commented on AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt:
```diff
@@ -596,28 +590,86 @@ open class DeckPicker :
+     *   relying instead on modifying it directly and/or using [onPrepareOptionsMenu].
+     *   Note an issue with the latter: https://github.com/ankidroid/Anki-Android/issues/7755
+     *
+     * TODO BEFORE-RELEASE Add tooltip text to the image button.
+     *   Menu items normally have titles that are shown on long tap.
+     *   As this menu item delegates the UI to the action view, it is up to that to handle presses.
+     *   When we decide on the text, use `TooltipCompat.setTooltipText` on the button to set it.
```
Ahh I left a TODO for the same and if I remember correctly there is an open PR which I don't think was updated after I mentioned the issue in that PR

criticalAY commented:
> @criticalAY tagging as you were requested as a reviewer in Discord

Done!

criticalAY commented:
@oakkitten @david-allison I just want to be sure if the menu is being refreshed after the migration completes as in my emulator it did not otherwise everything looks good, Thankyou!

oakkitten commented:
Updated to address suggestions and:
* resolve the potential division by 0 error (shame on me, I complained about this very issue in preceding PRs but totally missed it when I dealt with it myself 🤦‍♀️);
* revert a string key change for a string that didn't change (sorry, forgot to change it the first time)
* use `android:gravity="center"` in `dialog_with_progress_indicator.xml`

BrayanDSO reviewed (changes requested):
Tested on a 29 emulator.

1. The progress circle here isn't correctly aligned:

![image](https://user-images.githubusercontent.com/69634269/232315115-cdee9692-92bc-41ab-b6e8-ad7eefe2d9e1.png)

2. The circle progress bar indicator only shows up if I force a menu recreation (e.g. by opening the navigation drawer)

https://user-images.githubusercontent.com/69634269/232315178-ad1bb3ea-2767-4dcf-8a50-4d82c522ec17.mp4

3. Perhaps wouldn't be better to drop the notification progress bar after it is completed? I think that this more visually distinct, which easily denotes to the user that something different has happened (i.e. the migration has been finished)

oakkitten commented:
Updated to resolve the issue of the progress circle being improperly aligned.

> The circle progress bar indicator only shows up if I force a menu recreation (e.g. by opening the navigation drawer)

Cannot reproduce. I only have two devices, on Android 11 (API 30) everything works fine and on Android 6.0.1 I'm running into #13620. PC too weak to run an emulator. Does the issue also appear on the commit on main from where this PR branch stems? Anyway, I'm not sure if I can do anything here since I can't reproduce.

> Perhaps wouldn't be better to drop the notification progress bar after it is completed? I think that this more visually distinct, which easily denotes to the user that something different has happened (i.e. the migration has been finished)

I don't know, I like that the progress bar stays. So that the user knows that it's the same notification. I was thinking about having a different icon for it though, like a start with an arrow for transferring and a star with a checkmark for done. If there is time

BrayanDSO reviewed (approved):
The circle bar not appearing issue didn't happen on an API 33 emulator. As the issue showed up on a version that doesn't need the migration (29), it won't matter. 

So, LGTM. Pretty good code here. All I could think of to change are already stated on the TODOs

criticalAY reviewed (approved):
Considering the TODOs I'm okay with this one

david-allison reviewed (approved):
Requested changes made, looks great!

snowtimeglass commented on AnkiDroid/src/main/res/values/01-core.xml:
```diff
@@ -238,11 +238,14 @@
     <string name="migration_successful_message">Migration successful</string>
-    <string name="migration_failed_message">Migration failed</string>
-    <string name="migration_transferred_size">Migrated %1$.0f MB of %2$.0f MB</string>
+    <string name="migration__failed">Migration failed: %s</string>
```
Excuse me, can I see  in some way a screen of the app where this string is shown? (Not the string without colon, but the one with a colon and a specifier.) I would like to upload its screenshot to Crowdin as a translation resource, but I got lost.

oakkitten replied:
You'll see that in the notification, I added a comment at some later point: https://github.com/ankidroid/Anki-Android/blob/f6b1382dea1dc83a6d4361cd55c9484be91682ff/AnkiDroid/src/main/res/values/01-core.xml#L263-L265

snowtimeglass replied:
Thank you for your time and work. I guess that I have to "fail" the migration to see the string in the notification, but I still don't know how to do that. I would appreciate it if you would tell me the way of reproduction or provide a screenshot itself.

Also, if you would please add to the comment some information about what specific strings will go into the specifier (`%s`), it would be more helpful, I think. (I'm a bit concerned about the appropriateness of the word order of the following current Vietnamese translation.)

Original: `Migration failed: %s`
Vietnamese: `Chuyển %s không thành công.`
https://crowdin.com/translate/ankidroid/7290/en-vi?filter=basic&value=0#6536199

oakkitten replied:
Currently the errors look like this. We should probably skip the exception name for the notification...

Yeah the Vietnamese translation looks like it's a bit off.

![Screenshot_20230428-121423_AnkiDroid](https://user-images.githubusercontent.com/1710718/235134134-9f293c8a-4986-4414-9bf4-6ff677aa7e43.png)
![Screenshot_20230428-121431_AnkiDroid](https://user-images.githubusercontent.com/1710718/235134151-d05e2ff0-66e0-4cd4-a633-6e139d9a7508.png)

snowtimeglass replied:
Thank you so much!
